README.md
# Readme

Node.js server code for Subrosa.

### Requirements

* Node.js
* MySQL-compatible server (recommended: MariaDB)
* Redis
* GraphicsMagick
* URL forwaded to port 3000 (nginx will do)

### Setup
1. Clone this repository. 
2. Run `npm install`
3. Import `structure.sql`
4. Set up `config.js` (copy and edit `config.sample.js`)
5. Forward `yourhostname/server/` to port 3000 where websockets is listening
6. `node main.js` or `nodejs main.js`

Connect to the server by setting a custom network in Subrosa client (bottom of start screen). You'll need to host the client locally.


news
[{"title": "v1.02 released", "content": "Calls will now automatically reconnect if disconnected due to a network issue. Fixed rare race condition involving adding contacts.", "time": 1420635196778}, {"title": "v1.01 released", "content": "Fixed incorrect displaying of 'am' when it should've been 'pm' (thanks Pat), removed offline relaying node.", "time": 1420336164374}, {"title": "v1.00 released", "content": "Subrosa is officially out of beta, with the server source code released. Added support for connecting to different networks. Happy holidays!", "time": 1418777192004}]


db.js
var mysql = require('mysql2');
var config = require("./config.js");

var sql = mysql.createPool({
    host: config.db.host,
    user: config.db.user,
    password: config.db.password,
    socketPath: '/var/run/mysqld/mysqld.sock',
    database: config.db.database,
    connectionLimit: 50,
    multipleStatements: true,
    trace: false
});

var noderedis = require('redis');
var redis = noderedis.createClient();

redis.on("error", function (err) {
	console.log("Redis Error " + err);
});

function sqlQuery(query, response){
    sql.getConnection(function(err, connection) {
		if(err){
			console.log(err);
		}
        connection.query(query, function(err, rows){
            if(response){
                response.apply(this, arguments);
            }
            connection.release();
        });;
    });
}

module.exports.sql = sql;
module.exports.sqlQuery = sqlQuery;
module.exports.escape = mysql.escape;
module.exports.redis = redis;


main.js
createExports();
var db = require("./db.js");
var captcha = require('./captcha.js');
var users = require("./users.js");
var buffer = require("./buffer.js");
var messaging = require("./messaging.js");
var http = require('http');
var util = require('util');
var fs = require("fs");
var WebSocketServer = require('ws').Server;

var wss = new WebSocketServer({port: 4000});

var connections = [];
var uidConn = {};
var _usernameCache = {};
var setConnProto = false;

wss.on("connection", function(conn){
	if(!setConnProto){
		conn.constructor.prototype.msg = function(type, message){ if(this.readyState == this.OPEN){ message.sockType=type; this.send(JSON.stringify(message)) } }; 
		setConnProto=true;
	}
	conn.heartbeat = true;
	connections.push(conn);
	conn.on("message", function(message){
		if(conn.ignored)
			return;
		if(message.length>0){
			if(message[0] == "{"){
				try {
					message = JSON.parse(message);
				} catch ( error ) {
					console.log("Invalid JSON message received:" + error);
				}
				if(message.sockType){
					if(module.exports.map[message.sockType]){
						module.exports.map[message.sockType](conn, message);
					} else {
						console.log("Unrecognized mapping: " + message.sockType);
					}
				} else {
					console.log("No message type specified:" + message.sockType);
				}
			} else {
				// herbeat response
				conn.heartbeat = true;
			}
		}
	});
	conn.on("close", function(message){
		connections.splice(connections.indexOf(conn), 1);
		if(conn.uid){
			db.redis.hset("user:" + conn.uid, "online",  0);
			
			var uR = module.exports.userRoomSockets[conn.uid];
			var rS = module.exports.roomSockets;
			var gCI = module.exports.groupCallInfo;
			
			for(var i in uR){ // for every room I'm in
				if(rS[uR[i]] && rS[uR[i]].indexOf(conn.uid) !== -1){
					rS[uR[i]].splice(rS[uR[i]].indexOf(conn.uid), 1); // remove from the room socket list
				}
				if(gCI[uR[i]] && gCI[uR[i]].users.indexOf(conn.uid) != -1){ // if in a group call
					gCI[uR[i]].users.splice(gCI[uR[i]].users.indexOf(conn.uid), 1); // remove from group call
					module.exports.emitTo(rS[uR[i]], "", "statusUpdate", {target: conn.uid, status: 0}); // send offline (parsed by clients as exiting group call)
				}
			}
			
			uR = [];
			
			for(var i in module.exports.lastList[conn.uid]){
				if(uidConn[module.exports.lastList[conn.uid][i]]){ // send offline to contacts
					uidConn[module.exports.lastList[conn.uid][i]].msg("statusUpdate", {target: conn.uid, status: 0});
				}
			}
			
			uidConn[conn.uid] = null;
		}
	});
});

setInterval(function(){
	for(var i = 0; i < connections.length; i++){
		if(connections[i] && connections[i].send){
			connections[i].heartbeat = false;
			try {
				connections[i].send("*"); // send hearbeat
			} catch (error) {
				console.log(error);
			}
		}
	}
	setTimeout(function(){
		connections.filter(function(conn){
			if(!conn.heartbeat){
				conn.close();
				return false;
			}
			return true;
		});
	}, 20 * 1000);
}, 40 * 1000);

var lastVersionCheck = -1;
var lastVersionData = "";
module.exports.on("version", function(conn, data){
	if(new Date().getTime() > lastVersionCheck + (60 * 1000)){
		fs.readFile("version", function(err, data){
			lastVersionData = JSON.parse(data);
			lastVersionData.time = new Date().getTime();
			lastVersionCheck = new Date().getTime();
			conn.msg("version", lastVersionData);
		});
	} else {
		lastVersionData.time = new Date().getTime();
		conn.msg("version", lastVersionData);
	}
});
var lastNewsCheck = -1;
var lastNewsData = "";
module.exports.on("news", function(conn, data){
	if(new Date().getTime() > lastNewsCheck + (60 * 1000)){
		fs.readFile("news", function(err, data){
			lastNewsData = JSON.parse(data);
			lastNewsCheck = new Date().getTime();
			conn.msg("news", {news: lastNewsData});
		});
	} else {
		conn.msg("news", {news: lastNewsData});
	}
});
function createExports(){
	module.exports.map = {};
	module.exports.on = function(type, callback){
		if(module.exports.map[type]){
			console.log("WARNING: Rebinding for " + type);
		}
		module.exports.map[type] = callback;
	}
	module.exports.sortUID = function(uid1, uid2){
		var sorted = [uid1, uid2].sort();
		return sorted[0] + "-" + sorted[1];
	}
	module.exports.setUID = function(conn){
		if(conn.uid){
			uidConn[conn.uid] = conn;
		}
	}
	module.exports.emitTo = function(toUIDarray, exceptUID, type, data){
		if(true){
			for(var i = 0; i < toUIDarray.length; i++){
				if(toUIDarray[i] !== exceptUID && uidConn[toUIDarray[i]]){
					uidConn[toUIDarray[i]].msg(type, data);
				}
			}
		} else {
			// different server, tbd
		}
	}
	module.exports.lastList = [];
	module.exports.roomSockets = [];
	module.exports.userRoomSockets = []; // a list of roomSockets a user is in, so they can be exited when user quits
	module.exports.getConn = function(uid){
		if(uidConn[uid]){
			return uidConn[uid];
		}
		return false;
	}
	module.exports.usernameCache = function(uid, callback){
		// returns [username, bgColor]
		if(_usernameCache[uid])
			return callback(_usernameCache[uid]);
		db.redis.hmget("user:" + uid, "username", "bgColor", function(err, res){
			if(res){
				callback(res);
				_usernameCache[uid] = res;
			} else {
				return callback(["", ""]);
			}
		});
	};
	module.exports.checkRank = function(roomid, uid, rankRequired, performingOnUID, callback){
		db.redis.hmget("conv:" + roomid, "users", "ranks", function(err, res){
			if(res[0].indexOf(uid) != -1){
				var ranks = JSON.parse(res[1]);
				if(!ranks[uid] || ranks[uid] < rankRequired)
					return callback(false);
				if(performingOnUID && ranks[uid] <= ranks[performingOnUID] && ranks[uid] != 9)
					return callback(false);
				callback(true);
			}
		});
	};
	module.exports.merge = function(obj1, obj2){ // merge two objects
		var obj3 = {};
		for (var attrname in obj1) { obj3[attrname] = obj1[attrname]; }
		for (var attrname in obj2) { obj3[attrname] = obj2[attrname]; }
		return obj3;
	}
	module.exports.armor = function(){
		// ^ to denote in-item seperator and & to denoate item seperator, at start
		// escape char: !
		// &1^themaindata^auxdata&1^another!^main!!data^auxda!&ta
		var output = "";
		for(var i = 0; i < arguments.length; i++) {
			output += arguments[i].toString().replace(/!/g, "!!").replace(/\^/g, "!^").replace(/\&/g, "!&") + "^";
		}
		return "&" + output.substr(0, output.length-1); // cut off trailing ^
	}
	module.exports.groupCallInfo = [];
	module.exports.unarmor = function(input){
		// designed to work directly from a mysql conv buffer
		var output = [];
		var il = input.length;
		var selecting = false;
		var currentObj = [];
		var currentItem = "";
		
		for(var i = 0; i < il; i++){
			if(!selecting){
				if(input[i] == "&" && (i == 0 || input[i-1] != "!")){
					selecting = true;
				}
			} else {
				if(input[i] == "!" && i != il-1){
					if(i != il-2){
						currentItem += input[i+1];
						i++;
					}
				} else if(input[i] == "^"){
					currentObj.push(currentItem);
					currentItem = "";
				} else if(input[i] == "&" || i == il -1){
					if(i == il - 1){
						currentItem += input[i];
					}
					currentObj.push(currentItem);
					currentItem = "";
					if(currentObj.length == 5){
						output.push(currentObj);
					} else {
						console.log("main.js 248: currentObj length is " + currentObj.length, currentObj);
					}
					currentObj = [];
				} else {
					currentItem += input[i];
				}
			}
		}
		return output;
	}
}
process.on('uncaughtException', function (err) {
  console.log('Caught exception: ', err, err.stack);
});


buffer.js
var db = require('./db.js');
var main = module.parent.exports;
var fs = require('fs');

main.on("getBuffer", function(conn, data){
	if(conn.uid && data && data.target){
		db.redis.hget("conv:" + data.target, "users", function(err, res){
			if((data.target.length == 37 && data.target.indexOf(conn.uid) != -1) || res.indexOf(conn.uid) != -1){
				getConv(data.target, conn, true, data.small);
			} else {
				conn.msg("commBundle", {bundle: [], target: data.target, small: false});
			}
		});
	}
});

var commBufferNext = []; // Added in next swoop
var commBufferPending = []; // Pending to be added
var commBufferNextBusy = [];
function insertConv(target, type, data, auxdata, senderuid, time){
	var armored = main.armor(type, data, auxdata, senderuid, time);
	var armoredLength = armored.length;
	
	if(!commBufferNext[target]){
		commBufferNext[target] = [armored, armoredLength];
		commBufferNextBusy[target] = false;
	} else {
		commBufferNext[target][0] += armored;
		commBufferNext[target][1] += armoredLength;
	}
	if(!commBufferNextBusy[target]){
		processNextCommBuffer(target);
	}
}
module.exports.insertConv = insertConv;
function processNextCommBuffer(target){
	// Cleanup for previous calls
	commBufferNextBusy[target] = false;
	commBufferPending[target] = "";
	
	if(commBufferNext[target] && commBufferNext[target].length && commBufferNext[target][1] > 0){
		commBufferNextBusy[target] = true;
		
		// Move to pending
		commBufferPending[target] = commBufferNext[target][0];
		
		var combinedArmored = new Buffer(commBufferNext[target][0], "binary").toString("hex"); // converted to hex
		var combinedArmoredLength = commBufferNext[target][1];
		
		commBufferNext[target] = ["", 0]; // Clear next
		
		db.sqlQuery("SELECT LENGTH(blobbuffer) as bufferLength FROM conv WHERE id=" + db.escape(target), function(err, rows){
			if(rows.length == 0 && combinedArmoredLength < (64 * 1024)){
				// create new row
				db.sqlQuery("INSERT INTO conv(id, blobbuffer) VALUES(" + db.escape(target) + ", 0x" + combinedArmored + ")", function(err, rows){
					processNextCommBuffer(target);
				});
				
			} else {
				if(parseInt(rows[0].bufferLength) + parseInt(combinedArmoredLength) > (64 * 1024)){
					// need to overwrite earlier history
					var cutFront = (50 * 1024) + combinedArmoredLength;
					db.sqlQuery("UPDATE conv SET blobbuffer=CONCAT(SUBSTR(blobbuffer, -" + cutFront + "), 0x" + combinedArmored + ") WHERE id=" + db.escape(target), function(err, rows){
						processNextCommBuffer(target);
					});
				} else {
					// simply append
					db.sqlQuery("UPDATE conv SET blobbuffer=CONCAT(blobbuffer, 0x" + combinedArmored + ") WHERE id=" + db.escape(target), function(err, rows){
						processNextCommBuffer(target);
					});
				}
			}
		});
	}
}
// probably scope issues
function getConv(conv, conn, wantKeyExchange, small){
	var theQuery = "SELECT blobbuffer FROM conv WHERE id=" + db.escape(conv);
	if(small){ // only return last 4kb
		theQuery = "SELECT RIGHT(blobbuffer, " + (4 * 1024) + ") AS blobbuffer FROM conv WHERE id=" + db.escape(conv);
	}
	db.redis.hget("history:" + conn.uid, conv, function(err, historyRes){
		db.sqlQuery(theQuery, function(err, rows){
			var blob = "";
			if(rows[0] && rows[0].blobbuffer.length){
				blob += rows[0].blobbuffer.toString("binary");
			}
			if(commBufferPending[conv]){
				blob += commBufferPending[conv];
			}
			if(commBufferNext[conv] && commBufferNext[conv][0]){
				blob += commBufferNext[conv][0];
			}
			if(blob.length > 0){
				if(blob.length < 4 * 1024)
					small = false; // we don't have a larger buffer, so this is large.
				
				var unarmored = main.unarmor(blob);
				var unarmoredLength = unarmored.length;
				var usernameLookupsLeft = (conv.length == 20 ? unarmoredLength : 0);
				var sendBuffer = [];
				
				for(var i = 0; i < unarmored.length; i++){
					var messageType = unarmored[i][0];
					var messageData = unarmored[i][1];
					var messageAuxdata = unarmored[i][2];
					var senderUID = unarmored[i][3];
					var messageTime = unarmored[i][4];
					if(historyRes && historyRes >= messageTime){
						small = false;
						unarmoredLength--;
						if(usernameLookupsLeft)
							usernameLookupsLeft--;
						continue;
					}
					if(messageType == 1){
						if(wantKeyExchange){
							// delay sending until we get the public key
							new function(){
								var unshifted = {type: messageType, data: messageData, auxdata: messageAuxdata, sender: senderUID, time: messageTime};
								var searchUID = (conv.substr(4).split("-")[0] == conn.uid ? conv.substr(4).split("-")[1] : conv.substr(4).split("-")[0]);
								db.sqlQuery("SELECT publicKey FROM users WHERE uid='" + searchUID + "'", function(err, rows){
									if(rows && rows[0] && rows[0].publicKey){
										unshifted.pubKey = rows[0].publicKey;
										sendBuffer.unshift(unshifted);
										if(sendBuffer.length == unarmoredLength){
											// okay, let's send!
											conn.msg("commBundle", {bundle: sendBuffer, target: conv, small: small});
											sendBuffer = [];
										}
									}
								});
							}();
						} else {
							unarmoredLength--;
						}
					} else {
						var closureClosure = function(){
							var pushed = sendBuffer.push({type: messageType, data: messageData, auxdata: messageAuxdata, sender: senderUID, time: messageTime});
							if(conv.length==20){
								main.usernameCache(senderUID, function(username){
									sendBuffer[pushed-1].username = username[0];
									sendBuffer[pushed-1].bgColor = username[1];
									usernameLookupsLeft--;
									if(usernameLookupsLeft==0){
										conn.msg("commBundle", {bundle: sendBuffer, target: conv, small: small});
										sendBuffer = [];
									}
								});
							} else {
								if(sendBuffer.length == unarmoredLength){
									conn.msg("commBundle", {bundle: sendBuffer, target: conv, small: small});
									sendBuffer = [];
								}
							}
						}();
					}
				}
				if(sendBuffer.length == unarmoredLength && conv.length == 37){
					conn.msg("commBundle", {bundle: sendBuffer, target: conv, small: small});
				}
			} else {
				setTimeout(function(){
					conn.msg("commBundle", {bundle: [], target: conv, small: false});
					// delay in case of addList so this does not arrive prior
				}, 250);
			}
		})
	});
}
module.exports.getConv = getConv;


captcha.js
var gm = require('gm');
var fs = require('fs');
var pool = [];
var main = module.parent.exports;
var ipLockout = {};

main.on("getCaptcha", function(conn, data){
	var challenge = Math.random().toString().substr(2);
	var captchaText = Math.round(100000 + Math.random() * 899999);
	pool[challenge] = captchaText;
	if(conn.upgradeReq.headers['x-forwarded-for']){
		if(!ipLockout[conn.upgradeReq.headers['x-forwarded-for']]){
			ipLockout[conn.upgradeReq.headers['x-forwarded-for']] = [1, new Date().getTime()];
		} else {
			if(ipLockout[conn.upgradeReq.headers['x-forwarded-for']][1] > new Date().getTime() - 15 * 1000){
				ipLockout[conn.upgradeReq.headers['x-forwarded-for']][0]++;
				ipLockout[conn.upgradeReq.headers['x-forwarded-for']][1] = new Date().getTime();
			} else {
				ipLockout[conn.upgradeReq.headers['x-forwarded-for']] = [1, new Date().getTime()];
			}
		}
		if(ipLockout[conn.upgradeReq.headers['x-forwarded-for']] > 6){
			captchaText = "TOOMANY";
		}
	}
	
	var captchaImg = gm(220, 80, "#ffffff");
	
	captchaImg.stroke("#000000", 2).fill("#ffffff");
	
	for(var i = 0; i < 6; i++){
		var x = Math.random()*220;
		var y = Math.random()*80;
		captchaImg.drawCircle(x, y, x, y+10+Math.random()*30);
	}
	
	
	captchaImg.stroke("#000000", 1).fontSize(22+Math.floor(Math.random()*7)).fill("#000000").draw("translate 0, 0 rotate " + Math.round((-10+Math.random()*20)) + " text " + (25+Math.random()*75) + " " + (30+Math.random()*35) + "\"" + captchaText + "\"");
	
	
	captchaImg.quality(35).write("./tmp/captcha-" + challenge + ".jpg", function(err){
		if(err){
			console.log("Write captcha error " + err);
			return;
		}
		fs.readFile("./tmp/captcha-" + challenge + ".jpg", function(err, filedata){
			if(err){
				console.log("Read file error while reading CAPTCHA " + err);
				return;
			}
			conn.msg("captcha", {purpose: data.purpose, captcha: "data:image/jpeg;base64," + filedata.toString("base64"), challenge: challenge});
			fs.unlink("./tmp/captcha-" + challenge + ".jpg");
		});
	});
	
	setTimeout(function(){
		delete pool[challenge];
	}, 20 * 60 * 1000);	
});

module.exports.checkCaptcha = function(challenge, answer){
	if(pool[challenge]){
		if(pool[challenge] === parseInt(answer)){
			delete pool[challenge];
			return true;
		}
		delete pool[challenge];
	}
	return false;
}


users.js
var db = require('./db.js');
var main = module.parent.exports;
var gm = require('gm');
var fs = require('fs');
var exec = require('child_process').exec;
var captcha = require('./captcha.js');

var allowedBgColors = ["EDFBFF", "EDECFC", "F2E8FF", "F3DAFF", "FCE1F9", "FFECEF", "FFF5E8", "FFF9E8", "FEFFE8", "E8FDE5", "E9FFD2", "FCFDFC"];
main.on("userExists", function(conn, data){
	if(data.username){
		if(data.username.length<3){
			return conn.msg("userExists", {username: data.username, exists: false, error: "Too short."});
		}
		if(data.username.length>16){
			return conn.msg("userExists", {username: data.username, exists: false, error: "Too long."});
		}
		if(data.username.toLowerCase().indexOf("subrosa") != -1 || data.username.toLowerCase().indexOf("admin") != -1){
			return conn.msg("userExists", {username: data.username, exists: false, error: "Reserved."});
		}
		if(!data.username.match(/^[A-z0-9\-\.]*$/g)){
			return conn.msg("userExists", {username: data.username, exists: false, error: "Invalid characters."});
		}
		db.sqlQuery("SELECT uid FROM users WHERE username=" + db.escape(data.username), function(err, rows){
			conn.msg("userExists", {username: data.username, exists: (rows.length === 1), uid: (rows[0] ? rows[0].uid : "")});
		});
	}
});
main.on("register", function(conn, data){
	if(data.username && data.displayname && typeof data.email != 'undefined' && data.challenge && data.captcha && typeof data.newsletter != 'undefined' && data.derivedKeyHash && data.publicKey && data.encryptedBlob){
		if(data.username.length<3)
			return conn.msg("register", {status: "FAIL", message: "Username is too short.", restart: true});
		if(data.username.length>16)
			return conn.msg("register", {status: "FAIL", message: "Username is too long.", restart: true});
		if(!data.username.match(/^[A-z0-9\-\.]*$/g))
			return conn.msg("register", {status: "FAIL", message: "Username contains invalid characters.", restart: true});
		if(data.displayname.length<3)
			return conn.msg("register", {status: "FAIL", message: "Display name is too short.", restart: true});
		if(data.username.toLowerCase().indexOf("subrosa") != -1 || data.username.toLowerCase().indexOf("admin") != -1)
			return conn.msg("register", {status: "FAIL", message: "Username contains reserved characters.", restart: true});
		if(data.displayname.length>30)
			data.displayname = data.displayname.substr(0, 30);
		if(data.email.length<6 || data.email.indexOf("@") == -1 || data.email.indexOf(".") == -1)
			data.email = "email@skipped.com";
		if(!data.encryptedBlob.iv || !data.encryptedBlob.salt || !data.encryptedBlob.kdf || !data.encryptedBlob.data)
			return conn.msg("register", {status: "FAIL", message: "Something went wrong with your encryption key. Please try again.", restart: true});
		if(data.encryptedBlob.iv.length != 16 || data.encryptedBlob.salt.length != 32 || data.encryptedBlob.kdf.length != 3 || data.encryptedBlob.data.length < 1000 || data.derivedKeyHash.length != 64 || data.publicKey.length < 100)
			return conn.msg("register", {status: "FAIL", message: "Something went wrong with your encryption key.. Please try again.", restart: true});
		db.sqlQuery("SELECT COUNT(*) AS usernameUnique FROM users WHERE username=" + db.escape(data.username), function(err, rows){
			if(rows[0]['usernameUnique'] == 0){
				if(captcha.checkCaptcha(data.challenge, data.captcha)){
					var uid = randomUID();
					
					db.sqlQuery("INSERT INTO users(uid, username, displayname, email, newsletter, passwordhash, publickey, ivBin, saltBin, kdf, userblob, lastaccess) VALUES('" + uid + "', " + db.escape(data.username) + ", " + db.escape(data.displayname) + ", " + db.escape(data.email) + ", " + (data.newsletter?1:0) + ", " + db.escape(data.derivedKeyHash) + ", " + db.escape(data.publicKey) + ", 0x" + binaryStringToHex(data.encryptedBlob.iv) + ", 0x" + binaryStringToHex(data.encryptedBlob.salt) + ", " + db.escape(data.encryptedBlob.kdf) + ", " + db.escape(data.encryptedBlob.data) + ", NOW())", function(err, rows){
						if(err)
							return console.log("DB Error when registering: " + err);
							
						
						db.redis.hset("user:" + uid, "username", data.username);
						db.redis.hset("user:" + uid, "displayname", data.displayname);
						db.redis.hset("user:" + uid, "status",  1);
						db.redis.hset("user:" + uid, "bgColor", allowedBgColors[Math.floor(Math.random()*allowedBgColors.length)]);
						genAvatar(uid);
						return conn.msg("register", {status: "OK"});
					});
				} else {
					return conn.msg("register", {status: "FAIL", id: "CAPTCHA", message: "The CAPTCHA is incorrect. Please try again.", restart: false});
				}
			} else {
				return conn.msg("register", {status: "FAIL", message: "Someone took your username while you were registering! Try another.", restart: true});
			}
		});
	}
});
main.on("loginMain", function(conn, data){
	if(data.step){
		if(data.step == 1 && data.username){
			db.sqlQuery("SELECT HEX(saltBin) AS saltHex, kdf FROM users WHERE username=" + db.escape(data.username), function(err, rows){
				if(!rows || rows.length == 0){
					return conn.msg("loginMain", {step: 1, status: "FAIL", message: "Username does not exist."});
				} else {
					var salt = new Buffer(rows[0].saltHex, 'hex').toString('binary');
					return conn.msg("loginMain", {step: 1, status: "OK", salt: salt, kdf: rows[0].kdf});
				}
			});
		} else if(data.step == 2 && data.hash){
			db.sqlQuery("SELECT uid, passwordhash FROM users WHERE username=" + db.escape(data.username), function(err, rows){
				if(rows[0] && rows[0].passwordhash && rows[0].uid){
					setTimeout(function(){
						if(rows[0].passwordhash === data.hash){
							conn.uid = rows[0].uid;
							main.setUID(conn);
							db.redis.hset("user:" + conn.uid, "online", 1);
							
							main.userRoomSockets[conn.uid] = [];
							
							if(typeof data.resendBlob != 'undefined' && data.resendBlob == false)
								return main.map["getLists"](conn, {});

							db.sqlQuery("SELECT HEX(ivBin) AS ivHex, userblob, displayname FROM users WHERE username=" + db.escape(data.username), function(err, rows2){
								if(rows2[0] && rows2[0].ivHex && rows2[0].userblob){
									
									var iv = new Buffer(rows2[0].ivHex, 'hex').toString('binary');
									
									db.redis.hmget("user:" + conn.uid, "status", "avatar", "bgColor", function(err, res){
										conn.msg("loginMain", {step: 2, status: "OK", iv: iv, userBlob: rows2[0].userblob, uid: rows[0].uid, displayname: rows2[0].displayname, username: data.username, status: res[0], avatar: res[1], bgColor: res[2]});
										return main.map["getLists"](conn, {});
									});
								}
							});
						} else {
							return conn.msg("loginMain", {step: 2, status: "FAIL", message: "Password incorrect."});
						}
					}, 20);
				}
			});
		}
	}
});
main.on("updateBlob", function(conn, data){
	if(data.iv && data.blob && conn.uid){
		if(data.iv.length == 16 && data.blob.length > 500){
			db.sqlQuery("UPDATE users SET ivBin=0x" + binaryStringToHex(data.iv) + ", userblob=" + db.escape(data.blob) + " WHERE uid='" + conn.uid + "'", function(err, rows){
				if(err){
					return console.log("MySQL error while updating blob:" + err);
				}
			});
		}
	}
});
main.on("getPubKey", function(conn, data){
	if(conn.uid && data.uid){
		db.sqlQuery("SELECT publickey FROM users WHERE uid=" + db.escape(data.uid), function(err, rows){
			if(rows[0] && rows[0].publickey){
				conn.msg("getPubKey", {uid: data.uid, pubKey: rows[0].publickey});
			}
		});
	}
});
main.on("changeStatus", function(conn, data){
	if(conn.uid && data && data.status && typeof data.status == "number"){
		db.redis.hset("user:" + conn.uid, "status", data.status);
		for(var i in main.lastList[conn.uid]){
			main.emitTo([main.lastList[conn.uid][i]], "", "statusUpdate", {target: conn.uid, status: (data.status == 4 ? 0 : data.status)});
		}
	}
});
main.on("changeProfile", function(conn, data){
	if(conn.uid && data){
		if(data.displayname && data.displayname !== "*"){
			data.displayname = data.displayname.substr(0, 30);
			db.redis.hset("user:" + conn.uid, "displayname", data.displayname);
			db.sqlQuery("UPDATE users SET displayname=" + db.escape(data.displayname) + " WHERE uid='" + conn.uid + "'");
			for(var i in main.lastList[conn.uid]){
				main.emitTo([main.lastList[conn.uid][i]], "", "statusUpdate", {target: conn.uid, displayname: data.displayname});
			}
		}
		if(data.derivedKeyHash && data.derivedKeySalt && data.derivedKeyKdf){
				db.sqlQuery("UPDATE users SET passwordHash=" + db.escape(data.derivedKeyHash) + ", saltBin=0x" + binaryStringToHex(data.derivedKeySalt) + ", kdf=" + db.escape(data.derivedKeyKdf) + " WHERE uid='" + conn.uid + "'");
		}
		if(data.bgColor && typeof data.bgColor == "string"){
			if(allowedBgColors.indexOf(data.bgColor) != -1){
				db.redis.hset("user:" + conn.uid, "bgColor", data.bgColor);
			}
		}
	}
});
main.on("changeGroupInfo", function(conn, data){
	if(conn.uid && data){
		// check rank to be admin+
		db.redis.hmget("conv:" + data.target, "users", "ranks", function(err, res){
			if(!res)
				return;
			var ranks = JSON.parse(res[1]);
			if(res[0].indexOf(conn.uid) != -1 && ranks[conn.uid] >= 4){
				if(data.name){
					data.name = data.name.substr(0, 20);
					db.redis.hset("conv:" + data.target, "name", data.name);
					main.emitTo(main.roomSockets[data.target], conn.uid, "groupUpdate", {target: data.target, name: data.name});
				}
			}
		});
	}
});
main.on("filePart", function(conn, data){
	if(conn.uid && data && data.t && typeof data.p != 'undefined' && data.d){
		if(!conn.fileContinue)
			conn.fileContinue = "";
		if(data.target)
			conn.fileUploadTarget = data.target;
		if(data.d.length > 10240){
			return;
		} else if(data.p == 0){
			conn.fileContinue += data.d;
			if(conn.fileUploadTarget){
				if(conn.fileUploadTarget == "self"){
					setAvatar(conn, conn.uid, conn.fileContinue); 
					conn.fileContinue = "";
					conn.fileUploadTarget = "";
				} else if(conn.fileUploadTarget.length == 20){
					main.checkRank(conn.fileUploadTarget, conn.uid, 4, "", function(allowed){
						if(allowed){
							setAvatar(conn, conn.fileUploadTarget, conn.fileContinue);
							conn.fileContinue = "";
							conn.fileUploadTarget = "";
						}
					});
				}
			}
		} else {
			if(conn.fileContinue.length < 2048 * 1024){
				conn.fileContinue += data.d;
				conn.msg("gotPart", {});
			}
		}
	}
});
function setAvatar(conn, target, fileRaw){
	var tempFile = "./tmp/" + Math.round(Math.random()*9999999);
	var fileBuffer = new Buffer(fileRaw, "binary");
	fs.open(tempFile, "w", function(err, fd){
		fs.write(fd, fileBuffer, 0, fileBuffer.length, 0, function(err, written, buffer){
			db.sqlQuery("INSERT INTO avatars(uid, avatar) VALUES('" + target + "', 0x" + fileBuffer.toString("hex") + ")");
			gm(tempFile).resize(60, 60).quality(84).write(tempFile + ".jpg", function(err1){
				if(err1){
					console.log(err1);
					conn.msg("avatarSet", {error: true, target: (target == conn.uid ? "self" : target)});
					fs.unlink(tempFile + ".jpg");
					fs.unlink(tempFile);
				} else {
					fs.readFile(tempFile + ".jpg", function(err2, fileData){
						var encoded = "data:image/jpeg;base64," + fileData.toString("base64");
						conn.msg("avatarSet", {newAvatar: encoded, target: (target == conn.uid ? "self" : target)});
						fs.unlink(tempFile + ".jpg");
						fs.unlink(tempFile);
						
						if(target == conn.uid){
							db.redis.hset("user:" + target, "avatar", encoded);
							for(var i in main.lastList[conn.uid]){
								main.emitTo([main.lastList[conn.uid][i]], "", "avatar", {uid: conn.uid, avatar: encoded});
							}
						} else {
							db.redis.hset("conv:" + target, "avatar", encoded);
							main.emitTo(main.roomSockets[target], conn.uid, "avatar", {id: target, avatar: encoded});
						}
					});
				}
			});
		});
	});
}
function genAvatar(uid){
	var colors = ["2fbe1e", "279a19", "eecf27", "21b8db", "321ae1", "e11ade", "f85757", "1d5495", "f689ea", "a82f0f", "41fff2", "ff7641"];
	// replace with a better random avatar generation
	exec("gm convert -size 180x180 plasma: -filter Point -crop 60x60+60+60 -quality 75 ./tmp/avatar.jpg", function(err1){
		if(err1)
			console.log(err1);
		fs.readFile("./tmp/avatar.jpg", function(err2, fileData){
			var encoded = "data:image/jpeg;base64," + fileData.toString("base64");
			db.redis.hset("user:" + uid, "avatar", encoded);
		});
	});
}
function randomColor(colorArray){
	return colorArray.splice(Math.floor(Math.random()*colorArray.length, 1));
}
function randomUID(){
    var s= '';
    var randomchar=function(){
    	var n= Math.floor(Math.random()*62);
    	if(n<10) return n; //1-10
    	if(n<36) return String.fromCharCode(n+55); //A-Z
    	return String.fromCharCode(n+61); //a-z
    }
    while(s.length< 16) s+= randomchar();
    return s;
}
function binaryStringToHex(binaryString){
	return new Buffer(binaryString, 'binary').toString('hex');
}
main.on("errorReport", function(conn, data){
	console.log("** Client err " + new Date().toString(), data.trace, "UID: " + (conn.uid ? conn.uid : " none"));
});
